File: src/python/offer/38.字符串的排列.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
@File    :   38.字符串的排列.py
@Time    :   2020/09/03 20:07:27
@Author  :   wylu
@Version :   1.0
@Contact :   15wylu@gmail.com
@License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
@Desc    :
"""
from typing import List


class Solution:
    def permutation(self, s: str) -> List[str]:
        def dfs(c: int) -> None:
            if c == n:
                ans.append(''.join(s))
                return

            used = set()
            for i in range(c, n):
                if s[i] in used:
                    continue
                used.add(s[i])
                s[c], s[i] = s[i], s[c]
                dfs(c + 1)
                s[c], s[i] = s[i], s[c]

        ans = []
        s, n = list(s), len(s)
        dfs(0)
        return ans


if __name__ == '__main__':
    solu = Solution()
    print(solu.permutation('abc'))
    print(solu.permutation('aac'))

File: src/golang/offer/38.字符串的排列.go
package offer

/**
 * @File    :   38.字符串的排列.go
 * @Time    :   2021/06/22 08:53:20
 * @Author  :   wylu
 * @Version :   1.0
 * @Contact :   15wylu@gmail.com
 * @License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
 * @Desc    :
 */

func permutation(s string) []string {
	n := len(s)
	ans := []string{}
	seq := []byte(s)

	var dfs func(cur int)
	dfs = func(cur int) {
		if cur == n {
			ans = append(ans, string(seq))
			return
		}
		seen := map[byte]bool{}
		for i := cur; i < n; i++ {
			if seen[seq[i]] {
				continue
			}
			seen[seq[i]] = true
			seq[cur], seq[i] = seq[i], seq[cur]
			dfs(cur + 1)
			seq[cur], seq[i] = seq[i], seq[cur]
		}
	}

	dfs(0)
	return ans
}
